objects/src/accounts/code.rs

```diff
@@ -1,50 +1,67 @@
 ///
 /// Account's public interface consists of a set of account methods, each method being a Miden VM
 /// program. Thus, MAST root of each method commits to the underlying program. We commit to the
 /// entire account interface by building a simple Merkle tree out of all method MAST roots.
-#[derive(Debug, Clone, PartialEq, Eq)]
+#[derive(Debug, Clone)] // TODO: Add PartialEq, Eq
 pub struct AccountCode {
-    method_tree: MerkleTree,
+    method_tree: SimpleSmt,
     module: ModuleAst,
-    // methods: Vec<Program>, commented out because Program doesn't currently implement Eq and
-    // PartialEq. Also, there might be a better way of describing a set of programs as they
-    // might share a lot of common code blocks. In a way, we want something like a Program
-    // struct but with many entry points.
+    procedures: Vec<Digest>,
 }
 
 impl AccountCode {
+    // CONSTANTS
+    // --------------------------------------------------------------------------------------------
+    pub const ACCOUNT_CODE_PATH: &'static str = "context::account";
+
     // CONSTRUCTOR
     // --------------------------------------------------------------------------------------------
     /// Creates and returns a new definition of an account's interface compiled from the specified
     /// source code.
     pub fn new(source: &str) -> Result<Self, AccountError> {
-        let _module_ast = ModuleAst::parse(source)?;
+        let module_ast = ModuleAst::parse(source)?;
+        let module =
+            Module::new(LibraryPath::new(Self::ACCOUNT_CODE_PATH).expect("valid path"), module_ast);
 
-        // TODO: compile module AST into a set of program MASTs. To do this we need to expose
-        // a new method on the assembler to compile a module rather than a program (something
-        // similar to Assembler::compile_module() but without internal parameters).
-        //
-        // Open question: how to initialize the assembler? Specifically, which libraries to
-        // initialize it with. stdlib and midenlib are the two libraries we need for sure - but
-        // how to handle accounts which rely on some user-defined libraries? i.e., should there
-        // be a way to specify an "on-chain" library somehow?
+        let assembler = Assembler::default()
+            .with_library(&MidenLib::default())
+            .expect("failed to load miden-lib")
+            .with_library(&StdLibrary::default())
+            .expect("failed to load std-lib");
 
-        // TODO: build a Merkle tree out of MAST roots of compiled programs. The roots should
-        // be sorted so that the same set of programs always resolves to the same root. If the
-        // number of programs is not a power of two, the remaining leaves should be set to ZEROs.
+        let mut procedure_digests = assembler
+            .compile_module(&module, &mut AssemblyContext::default())
+            .map_err(AccountError::AccountCodeAsselmberError)?;
+        procedure_digests.sort();
```

Comment:
Unfortunately, this sort will probably not yield good results as it is done based on Montgomery representation of field elements (and is probably sorting in "big-endian" order based on that). See note [here](https://github.com/0xPolygonMiden/crypto/blob/next/src/hash/rpo/digest.rs#L132). If we want the order to be "portable", we should either update the `Ord` implementation for `RpoDigest` or use a different procedure here. I'm leaning toward the latter option.

Reply:
What do you mean by the order being "portable"?  For the purpose of ordering account procedure I don't think the way in which we order is particularly important as its not enforced on the protocol level (maybe it should be?).  The only requirement we have here is that the ordering is consistent and deterministic.  Having said that if there is a more natural way of ordering then we should go with that.  Could you provide some guidance around how you would propose we implement a different procedure for ordering please?

Reply:
By portable I mean that we don't have to make assumptions about internal implementation of `Felt`'s. So, if some other library implements `Felt`'s not using Montgomery form, sorting would still be consistent.

A better way to sort would be using canonical representation of field elements. We can get that via `Felt::as_int()` method. So, basically, here we convert `[Felt; 4]` into `[u64; 4]` and then sort based on this (probably assuming little-endian order?).

Reply:
Ok that makes sense.  How about converting `[Felt; 4]` to `[u8; 32]` using `as_bytes()` and then sorting based on this?

https://github.com/0xPolygonMiden/crypto/blob/09025b40148f475d0c79e075d3c49bea32e574ea/src/hash/rpo/digest.rs#L34-L45

```rust
procedure_digests.sort_by(|a, b| a.as_bytes().cmp(&b.as_bytes()));
```

Reply:
Yep - that works!